cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -184,18 +280,51 @@ class InputState {
     return queue_.UnsyncFront();
   }
 
+#define LATEST_VAL_CASE(id, val)                            \
+  case Type::id: {                                          \
+    using T = typename TypeIdTraits<Type::id>::Type;        \
+    using CType = typename TypeTraits<T>::CType;            \
+    return val(data->GetValues<CType>(1)[latest_ref_row_]); \
+  }
+
   KeyType GetLatestKey() const {
-    return queue_.UnsyncFront()
-        ->column_data(key_col_index_)
-        ->GetValues<KeyType>(1)[latest_ref_row_];
+    if (key_hasher_ != NULLPTR) {
```

Comment:
When do you use key_hasher and when do you use the original value?

Reply:
The key_hasher is (made available and) used when the faster-path, which computes using `norm_value` on a single-key, does not apply. Currently, this only happens with a multi-key, but in the future will also happen when the single-key type cannot be coerced into a 64-bits ordered type, e.g., with a string-typed single-key.